src/main/host/syscall/handler/timerfd.rs

```diff
@@ -0,0 +1,244 @@
+                return Err(Errno::EINVAL.into());
+            };
+
+            // this is a clippy false-positive
+            #[allow(clippy::let_and_return)]
+            let val = runnable.memory_borrow().read(new_value_ptr)?;
+            val
+        };
+
+        // Verify a valid range for new_time nanosecond vals.
+        if new_value.it_value.tv_nsec < 0
+            || new_value.it_value.tv_nsec >= 999_999_999
+            || new_value.it_interval.tv_nsec < 0
+            || new_value.it_interval.tv_nsec >= 999_999_999
+        {
+            return Err(Errno::EINVAL.into());
+        }
+
+        // First, write out the old time if requested.
+        if !old_value_ptr.is_null() {
+            // Old value is always relative, even if TFD_TIMER_ABSTIME is set.
+            Self::timerfd_gettime_helper(ctx, timerfd, old_value_ptr)?;
+        }
+
+        // Now we can adjust the timer with the new_value.
+        if new_value.it_value.tv_sec == 0 && new_value.it_value.tv_nsec == 0 {
+            // A value of 0 disarms the timer; it_interval is ignored.
+            TimerFd::disarm_timer(timerfd);
+            log::trace!("TimerFd {} disarmed", fd);
+        } else {
+            // Need to arm the timer, value may be absolute or relative.
+            let value: SimulationTime = new_value.it_value.try_into().unwrap();
+            let interval: SimulationTime = new_value.it_interval.try_into().unwrap();
+            let now = Worker::current_time().unwrap();
+
+            let expire_time = {
+                let base = match flags.contains(TimerSetTimeFlags::TFD_TIMER_ABSTIME) {
+                    true => EmulatedTime::UNIX_EPOCH,
+                    false => now,
+                };
+                // The man page does not specify what happens if the configured time is in the past.
+                // On Linux, the result is an immediate timer expiration.
+                EmulatedTime::max(base + value, now)
+            };
+
+            TimerFd::arm_timer(timerfd, ctx.objs.host, expire_time, interval);
+
+            log::trace!(
+                "TimerFd {} armed to expire at {:?} (in {:?})",
+                fd,
+                expire_time,
+                expire_time.duration_since(&now)
+            );
+        }
+
+        Ok(0.into())
+    }
+}
+
+/// Checks the clockid; returns `Ok(())` if the clockid is `CLOCK_REALTIME` or
+/// `CLOCK_MONOTONIC`, or the appropriate errno if the clockid is unknown or
+/// unsupported.
+fn check_clockid(clockid: libc::c_int) -> Result<(), SyscallError> {
+    if clockid == libc::CLOCK_MONOTONIC || clockid == libc::CLOCK_REALTIME {
+        Ok(())
+    } else if clockid == libc::CLOCK_BOOTTIME
+        || clockid == libc::CLOCK_BOOTTIME_ALARM
+        || clockid == libc::CLOCK_REALTIME_ALARM
+    {
+        log::debug!(
+            "Unsupported clockid {}, we support CLOCK_REALTIME and CLOCK_MONOTONIC.",
+            clockid
+        );
+        Err(Errno::ENOSYS.into())
```

Comment:
Optional nit: I think it would be better to return `EINVAL` here too. It can be confusing when the logs show `ENOSYS`, since then it seems like we don't support this syscall at all, and that we're maybe not intercepting it properly.